airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/common.py

```diff
@@ -0,0 +1,155 @@
+from pydantic import BaseModel, ValidationError
+from source_amazon_ads.common import PageToken, SourceContext
+from source_amazon_ads.schemas import JSModel
+
+URL_BASE = "https://advertising-api.amazon.com/"
+
+
+class ErrorResponse(BaseModel):
+    code: str
+    details: str
+    requestId: str
+
+
+class BasicAmazonAdsStream(Stream, ABC):
+    def __init__(self, config, context: SourceContext = None):
+        self._ctx = context or SourceContext()
+        self._config = config
+        self._url = self._config.host or URL_BASE
+
+    @property
+    @abstractmethod
+    def model(self) -> JSModel:
+        """
+        Pydantic model to represent json schema
+        """
+
+    def get_json_schema(self):
+        return self.model.schema()
+
+
+# Basic full refresh stream
+class AmazonAdsStream(HttpStream, BasicAmazonAdsStream):
+    def __init__(self, config, *args, context: SourceContext = None, **kwargs):
+        BasicAmazonAdsStream.__init__(self, config, context=context or SourceContext())
+        HttpStream.__init__(self, *args, **kwargs)
+
+    @property
+    def url_base(self):
+        return self._url
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_headers(self, *args, **kvargs) -> MutableMapping[str, Any]:
+        return {"Amazon-Advertising-API-ClientId": self._config.client_id}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an object representing single record in the response
+        """
+        yield from json.loads(response.text) if response else []
+
+    def _send_request(self, request: requests.PreparedRequest, request_kwargs: Mapping[str, Any]) -> requests.Response:
+        try:
+            return super()._send_request(request, request_kwargs)
+        except requests.exceptions.HTTPError as http_exception:
+            response = http_exception.response
+            if response.status_code in [HTTPStatus.FORBIDDEN]:
+                try:
+                    resp = ErrorResponse.parse_raw(response.text)
+                    self.logger.warn(
+                        f"Unexpected error {resp.code} when processing request {request.url} for {request.headers['Amazon-Advertising-API-Scope']} profile: {resp.details}"
+                    )
+                except ValidationError:
+                    raise http_exception
+            else:
+                raise http_exception
+
+
+class ContextStream(AmazonAdsStream):
+    """
+    Stream for getting resources based on context set by previous stream.
+    """
+
+    flattern_properties = []
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        for record in super().parse_response(response, **kwargs):
+            for prop in self.flattern_properties:
+                if prop in record:
+                    record[prop] = json.dumps(record[prop])
```

Comment:
Why should we do `json.dump`s here?

Reply:
To convert complex json objects like [sponsored products campagin's](https://advertising.amazon.com/API/docs/en-us/sponsored-brands/3-0/openapi#/Campaigns/listCampaigns) landingPage to plain string. It would make json schema more simple and avoid creating crazy amount of  entangled database tables for relative db destination connectors  with normalization option